<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="&gt; Efficient operations for single cell ATAC-seq fragments and
    RNA counts matrices. Interoperable with standard file formats, and introduces
    efficient bit-packed formats that allow large storage savings and increased
    read speeds.">
<title>Single Cell Counts Matrices to PCA â€¢ BPCells</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Single Cell Counts Matrices to PCA">
<meta property="og:description" content="&gt; Efficient operations for single cell ATAC-seq fragments and
    RNA counts matrices. Interoperable with standard file formats, and introduces
    efficient bit-packed formats that allow large storage savings and increased
    read speeds.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="benparks.net" src="https://plausible.benparks.net/js/visit-counts.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">BPCells</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/pbmc3k.html">Basic tutorial</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bnprks/BPCells/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="bpcells">BPCells<a class="anchor" aria-label="anchor" href="#bpcells"></a>
</h1></div>
<p>BPCells is a package for high performance single cell analysis. It is designed to cover the processing pipeline from ATAC fragments or RNA counts matrices through to normalization, basic QC, and PCA.</p>
<p>Three of the key distinguishing features that allow for high performance in BPCells are:</p>
<ol style="list-style-type: decimal">
<li>Bit-packing compression to allow for extremely compact storage of fragments or counts matrices on-disk or in memory.</li>
<li>C++ code that operates on all data in a streaming fashion to support low memory usage and efficient use of CPU cache.</li>
<li>Matrix-free SVD solvers in combination with implicit normalization calculations to support computing the PCA of a normalized matrix while only ever storing the original counts matrix in memory.</li>
</ol>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>BPCells is easiest to install directly from github:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"bnprks/BPCells"</span><span class="op">)</span></span></code></pre></div>
<p>Before installing, you must have the HDF5 library installed on your system. HDF5 can be installed from your choice of package manager:</p>
<ul>
<li>conda: <code>conda install -c anaconda hdf5</code>
</li>
<li>apt: <code>sudo apt-get install libhdf5-dev</code>
</li>
<li>yum: <code>yum install hdf5-devel</code>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>Two key principles to understand about using BPCells is that all operations are <em>streaming</em> and <em>lazy</em>.</p>
<p>Streaming means that only a minimal amount of data is stored in memory while a computation is happening. There is almost no memory used storing intermediate results. Hence, we can compute operations on large matrices without ever loading them into memory.</p>
<p>Lazy means that no real work is performed on matrix or fragment objects until the result needs to be returned as an R object or written to disk. This helps support the streaming computation, since otherwise we would be forced to compute intermediate results and use additional memory.</p>
<div class="section level3">
<h3 id="basic-usage">Basic usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h3>
<p>We begin with a basic example of loading ATAC fragments from a 10x fragments file, reading a peak set from a bed file, then calculating a cell x peak matrix.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bnprks.github.io/BPCells">"BPCells"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># File reading is lazy, so this is instantaneous</span></span>
<span><span class="va">fragments</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/open_fragments_10x.html">open_fragments_10x</a></span><span class="op">(</span><span class="st">"atac_fragments.tsv.gz"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># This is when we actually read the file, should take 1-2 minutes to scan</span></span>
<span><span class="co"># since we bottleneck on gzip decompression.</span></span>
<span><span class="va">packed_fragments</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/fragment_io.html">write_fragments_memory</a></span><span class="op">(</span><span class="va">fragments</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Important to set compress=FALSE for speed. Should take a few seconds</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">packed_fragments</span>, <span class="st">"fragments.rds"</span>, compress<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Reloading from disk is only a few seconds now.</span></span>
<span><span class="va">packed_fragments</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"fragments.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/read_bed.html">read_bed</a></span><span class="op">(</span><span class="st">"peaks.bed"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># This is fast because the peak matrix calculation is lazy.</span></span>
<span><span class="co"># It will be computed on-the-fly when we actually need results from it.</span></span>
<span><span class="va">peak_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/peak_matrix.html">peak_matrix</a></span><span class="op">(</span><span class="va">packed_fragments</span>, <span class="va">peaks</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Here is where the peak matrix calculation happens. Should take</span></span>
<span><span class="co"># under 10 seconds.</span></span>
<span><span class="va">R_matrix</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="va">peak_matrix</span>, <span class="st">"dgCMatrix"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="streaming-operations">Streaming operations<a class="anchor" aria-label="anchor" href="#streaming-operations"></a>
</h3>
<p>The lazy, stream-oriented design means that we can calculate more complicated transformations in a single pass. This is both faster and more memory-efficient than calculating several intermediate results in a sequential manner.</p>
<p>As an example, we will perform the following pipeline: 1. Exclude fragments from non-standard chromosomes 2. Subset our cells 3. Add a Tn5 offset 4. Calculate the peak matrix 5. Calculate the mean-accessibility per peak</p>
<p>If this were done using e.g.Â GRanges or sparse matrices, we would need to do 3 passes through the fragments while saving intermediate results, and 2 passes over the peak matrix.</p>
<p>With BPCellâ€™s streaming operations, this can all be done directly from the fragments in a single pass, and the memory usage is limited to a few bytes per cell for iterating over the peak matrix and returning the colMeans.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Here I make use of the pipe operator (%&gt;%) for better readability</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://tidyverse.tidyverse.org" class="external-link">"tidyverse"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># We'll subset to just the standard chromosomes</span></span>
<span><span class="va">standard_chr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span></span>
<span>  <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="reference/IterableFragments-methods.html">chrNames</a></span><span class="op">(</span><span class="va">packed_fragments</span><span class="op">)</span>, <span class="st">"^chr[0-9XY]+$"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Pick a random subset of 100 cells to consider</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1337</span><span class="op">)</span></span>
<span><span class="va">keeper_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="reference/IterableFragments-methods.html">cellNames</a></span><span class="op">(</span><span class="va">packed_fragments</span><span class="op">)</span>, <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run the pipeline, and save the average accessibility per peak</span></span>
<span><span class="va">peak_accessibility</span> <span class="op">&lt;-</span> <span class="va">packed_fragments</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/select_chromosomes.html">select_chromosomes</a></span><span class="op">(</span><span class="va">standard_chr</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/select_cells.html">select_cells</a></span><span class="op">(</span><span class="va">keeper_cells</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/shift_fragments.html">shift_fragments</a></span><span class="op">(</span>shift_start<span class="op">=</span><span class="fl">4</span>, shift_end<span class="op">=</span><span class="op">-</span><span class="fl">5</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/peak_matrix.html">peak_matrix</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Note that if we knew the cell names ahead of time, we could even perform this operation directly on our orignal 10x fragments without ever saving the fragments into memory. This would be fairly slow because 10x fragment files are slow to decompress. With upcoming support for storing packed fragments directly on disk, this can become a much faster operation without ever needing to store fragments in memory.</p>
</div>
</div>
<div class="section level2">
<h2 id="roadmap">Roadmap<a class="anchor" aria-label="anchor" href="#roadmap"></a>
</h2>
<div class="section level3">
<h3 id="current-support">Current support:<a class="anchor" aria-label="anchor" href="#current-support"></a>
</h3>
<ul>
<li>Fragments
<ul>
<li>Reading/writing 10x fragment files on disk</li>
<li>Reading/writing packed fragment objects in memory or directly from disk</li>
<li>Interconversion of fragments objects with GRanges</li>
<li>Calculation of Cell x Peak matrices</li>
</ul>
</li>
<li>Matrices
<ul>
<li>Conversion to/from R sparse matrices</li>
<li>Read-write access to 10x hdf5 feature matrices, and read-only access to AnnData files, and</li>
<li>Reading/writing of packed sparse matrices in memory or directly from disk</li>
<li>Multiplication by dense matrices or vectors</li>
<li>Calculation of statistics like rowSums, colSums, rowMeans, and colMeans</li>
<li>Transparent handling of vector <code>+</code>, <code>-</code>, <code>*</code>, <code>/</code>, and <code>log1p</code> for streaming normalization. This allows implementation of ATAC-seq LSI and Seurat default normalization.</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="upcoming-additions">Upcoming additions:<a class="anchor" aria-label="anchor" href="#upcoming-additions"></a>
</h3>
<ul>
<li>Support for additional fragment formats:
<ul>
<li>Read fragments from bam files</li>
<li>Support direct download of files from URLs</li>
</ul>
</li>
<li>Support for additional matrix formats:
<ul>
<li>Write hdf5 AnnData matrices</li>
</ul>
</li>
<li>Support for additional matrix normalizations:
<ul>
<li>sctransform normalization</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="performance-estimates">Performance estimates<a class="anchor" aria-label="anchor" href="#performance-estimates"></a>
</h3>
<ul>
<li>Bit-packed storage:
<ul>
<li><p>Packed fragments are about 2x smaller than ArchR fragments and gzipped 10x fragment files</p></li>
<li><p>Packed fragments can be decompressed at &gt;5GB/s, and so decoding is disk-limited on SSDs or RAID arrays while remaining competitive with reading directly from uncompressed RAM</p></li>
<li>
<p>Packed matrices are 4-6x smaller than the equivalent sparse matrices, with similarly excellent decompression speed.</p>
<p>(Note: compared to R dgCMatrices, the numbers are more like 6-8x smaller due to dgCMatrices always storing double-precision floats)</p>
</li>
</ul>
</li>
<li>Comparison to ArchR:
<ul>
<li>Fragment import speed is about 10-fold faster</li>
<li>Peak and tile matrix calculations are about 10-fold faster than ArchR</li>
<li>BPCells usually uses less memory than ArchR, though ArchRâ€™s memory usage is never excessive</li>
</ul>
</li>
<li>PCA calculation:
<ul>
<li>Compared to Seurat/Scanpyâ€™s default normalization + PCA, BPCells can be nearly 100x more memory efficient for large datasets (1M cells), even when subsetting to just 2,000 genes.</li>
<li>The time for PCA calculation is faster than Seurat, but about 2x slower than Scanpy (BPCells) takes about 5 minutes to run PCA on a 1M cell matrix</li>
<li>Compared to Seuratâ€™s default normalization + PCA, BPCells will likely be about 10x more efficient in memory and CPU. It is unclear if this will multiply with the 4-6x memory savings from bitpacking counts matrices.</li>
<li>LSI is not expected to be substantially faster than ArchR, although the C++ implementation may provide a several-fold speedup, and bitpacking will provide a 4-6x memory savings.</li>
</ul>
</li>
</ul>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/bnprks/BPCells/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing BPCells</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Benjamin Parks <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-0261-7472" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Benjamin Parks.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  <img src="https://plausible.benparks.net/flask-plausible/bpcells-docs.png" style="position:absolute;">
</body>
</html>
