<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ BPCells</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="benparks.net" src="https://plausible.benparks.net/js/plausible.js"></script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BPCells</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/bnprks/BPCells/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="fragment-i-o">Fragment I/O <a href="#fragment-i-o" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Read &amp; write fragment objects</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="open_fragments_10x.html">open_fragments_10x()</a></code> </p>
        </td>
        <td><p>Read a 10x fragments file</p></td>
      </tr><tr><td>
          <p><code><a href="open_fragments_dir.html">open_fragments_dir()</a></code> </p>
        </td>
        <td><p>Open an existing fragments object from a directory</p></td>
      </tr><tr><td>
          <p><code><a href="open_fragments_hdf5.html">open_fragments_hdf5()</a></code> </p>
        </td>
        <td><p>Read a fragments object from an hdf5 file</p></td>
      </tr><tr><td>
          <p><code><a href="write_fragments_10x.html">write_fragments_10x()</a></code> </p>
        </td>
        <td><p>Write to a 10x fragments file</p></td>
      </tr><tr><td>
          <p><code><a href="write_fragments_dir.html">write_fragments_dir()</a></code> </p>
        </td>
        <td><p>Make a fragments object in binary files within a directory.</p></td>
      </tr><tr><td>
          <p><code><a href="write_fragments_hdf5.html">write_fragments_hdf5()</a></code> </p>
        </td>
        <td><p>Write a fragments object in an hdf5 file</p></td>
      </tr><tr><td>
          <p><code><a href="write_fragments_memory.html">write_fragments_memory()</a></code> </p>
        </td>
        <td><p>Make a fragments object in memory.</p></td>
      </tr><tr><td>
          <p><code><a href="convert_to_fragments.html">convert_to_fragments()</a></code> </p>
        </td>
        <td><p>Build a Fragments object from an R data frame or GRanges object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="atac-analysis-utilities">ATAC Analysis Utilities <a href="#atac-analysis-utilities" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="qc_scATAC.html">qc_scATAC()</a></code> </p>
        </td>
        <td><p>Calculate ArchR-compatible per-cell QC statistics</p></td>
      </tr><tr><td>
          <p><code><a href="nucleosome_counts.html">nucleosome_counts()</a></code> </p>
        </td>
        <td><p>Count fragments by nucleosomal size</p></td>
      </tr><tr><td>
          <p><code><a href="footprint.html">footprint()</a></code> </p>
        </td>
        <td><p>Get footprints around a set of genomic coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="peakMatrix.html">peakMatrix()</a></code> </p>
        </td>
        <td><p>Calculate ranges x cells overlap matrix</p></td>
      </tr><tr><td>
          <p><code><a href="tileMatrix.html">tileMatrix()</a></code> </p>
        </td>
        <td><p>Calculate ranges x cells tile overlap matrix</p></td>
      </tr><tr><td>
          <p><code><a href="geneScoreTilesArchR.html">geneScoreTilesArchR()</a></code> </p>
        </td>
        <td><p>Calculate gene-tile distances for ArchR gene activities</p></td>
      </tr><tr><td>
          <p><code><a href="geneScoreWeightsArchR.html">geneScoreWeightsArchR()</a></code> </p>
        </td>
        <td><p>Calculate GeneActivityScore weights matrix</p></td>
      </tr><tr><td>
          <p><code><a href="length_distribution.html">length_distribution()</a></code> </p>
        </td>
        <td><p>Count fragments by size</p></td>
      </tr><tr><td>
          <p><code><a href="order_ranges.html">order_ranges()</a></code> </p>
        </td>
        <td><p>Get end-sorted ordering for genome ranges</p></td>
      </tr><tr><td>
          <p><code><a href="read_bed.html">read_bed()</a></code> </p>
        </td>
        <td><p>Read a bed file file into a data frame</p></td>
      </tr><tr><td>
          <p><code><a href="read_encode_blacklist.html">read_encode_blacklist()</a></code> </p>
        </td>
        <td><p>Download ENCODE blacklist annotations</p></td>
      </tr><tr><td>
          <p><code><a href="read_gencode_genes.html">read_gencode_genes()</a></code> </p>
        </td>
        <td><p>Download gene annotations from gencode</p></td>
      </tr><tr><td>
          <p><code><a href="read_gtf.html">read_gtf()</a></code> </p>
        </td>
        <td><p>Read a gtf file into a data frame</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="fragment-transformers">Fragment Transformers <a href="#fragment-transformers" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Transforming fragment objects</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="shift_fragments.html">shift_fragments()</a></code> </p>
        </td>
        <td><p>Shift start or end coordinates of fragments by a fixed amount</p></td>
      </tr><tr><td>
          <p><code><a href="select_chromosomes.html">select_chromosomes()</a></code> </p>
        </td>
        <td><p>Select chromosmes for subsetting or translating chromosome IDs in a fragments object</p></td>
      </tr><tr><td>
          <p><code><a href="select_cells.html">select_cells()</a></code> </p>
        </td>
        <td><p>Select cells for subsetting or translating cells IDs in a fragments object</p></td>
      </tr><tr><td>
          <p><code><a href="subset_lengths.html">subset_lengths()</a></code> </p>
        </td>
        <td><p>Subset fragments to only include those in a given size range</p></td>
      </tr><tr><td>
          <p><code><a href="selectRegions.html">selectRegions()</a></code> </p>
        </td>
        <td><p>Select fragments overlapping (or not overlapping) selected regions</p></td>
      </tr><tr><td>
          <p><code><a href="prefix_cell_names.html">prefix_cell_names()</a></code> </p>
        </td>
        <td><p>Rename cells by adding a prefix to the names (e.g. sample name)</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="fragment-utilities">Fragment Utilities <a href="#fragment-utilities" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Other fragment functions</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="IterableFragments-methods.html">show(<i>&lt;IterableFragments&gt;</i>)</a></code> <code><a href="IterableFragments-methods.html">cellNames()</a></code> <code><a href="IterableFragments-methods.html">`cellNames&lt;-`()</a></code> <code><a href="IterableFragments-methods.html">chrNames()</a></code> <code><a href="IterableFragments-methods.html">`chrNames&lt;-`()</a></code> </p>
        </td>
        <td><p>IterableFragments methods</p></td>
      </tr><tr><td>
          <p><code><a href="scan_fragments.html">scan_fragments()</a></code> </p>
        </td>
        <td><p>Scan through fragments without performing any operations (used for benchmarking)</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="matrix-i-o">Matrix I/O <a href="#matrix-i-o" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions for reading &amp; writing sparse matrices</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="open_matrix_10x_hdf5.html">open_matrix_10x_hdf5()</a></code> </p>
        </td>
        <td><p>Read a sparse integer matrix from a 10x formatted hdf5 feature matrix file</p></td>
      </tr><tr><td>
          <p><code><a href="open_matrix_anndata_hdf5.html">open_matrix_anndata_hdf5()</a></code> </p>
        </td>
        <td><p>Read a sparse integer matrix from an anndata matrix in an hdf5 file</p></td>
      </tr><tr><td>
          <p><code><a href="open_matrix_dir.html">open_matrix_dir()</a></code> </p>
        </td>
        <td><p>Open an existing integer sparse matrix object written with write_matrix_dir</p></td>
      </tr><tr><td>
          <p><code><a href="open_matrix_hdf5.html">open_matrix_hdf5()</a></code> </p>
        </td>
        <td><p>Read a sparse integer matrix from an hdf5 file</p></td>
      </tr><tr><td>
          <p><code><a href="write_matrix_10x_hdf5.html">write_matrix_10x_hdf5()</a></code> </p>
        </td>
        <td><p>Write a matrix to a 10x-compatible hdf5 feature matrix file</p></td>
      </tr><tr><td>
          <p><code><a href="write_matrix_dir.html">write_matrix_dir()</a></code> </p>
        </td>
        <td><p>Make an integer sparse matrix object in binary files within a directory</p></td>
      </tr><tr><td>
          <p><code><a href="write_matrix_hdf5.html">write_matrix_hdf5()</a></code> </p>
        </td>
        <td><p>Write a sparse integer matrix to an hdf5 file</p></td>
      </tr><tr><td>
          <p><code><a href="write_matrix_memory.html">write_matrix_memory()</a></code> </p>
        </td>
        <td><p>Write a sparse matrix object to memory.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="matrix-utilities">Matrix Utilities <a href="#matrix-utilities" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="matrixStats.html">matrixStats()</a></code> </p>
        </td>
        <td><p>Calculate matrix stats</p></td>
      </tr><tr><td>
          <p><code><a href="convert_matrix_type.html">convert_matrix_type()</a></code> </p>
        </td>
        <td><p>Convert the type of a matrix</p></td>
      </tr><tr><td>
          <p><code><a href="transpose_storage_order.html">transpose_storage_order()</a></code> </p>
        </td>
        <td><p>Transpose the storage order for a matrix</p></td>
      </tr><tr><td>
          <p><code><a href="IterableMatrix-methods.html">show(<i>&lt;IterableMatrix&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">t(<i>&lt;IterableMatrix&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">`%*%`(<i>&lt;IterableMatrix&gt;</i>,<i>&lt;matrix&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">rowSums(<i>&lt;IterableMatrix&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">colSums(<i>&lt;IterableMatrix&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">rowMeans(<i>&lt;IterableMatrix&gt;</i>)</a></code> <code><a href="IterableMatrix-methods.html">colMeans(<i>&lt;IterableMatrix&gt;</i>)</a></code> </p>
        </td>
        <td><p>IterableMatrix methods</p></td>
      </tr><tr><td>
          <p><code><a href="min_scalar.html">min_scalar()</a></code> </p>
        </td>
        <td><p>Take the elementwise min with a constant</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="internal-functions">Internal Functions <a href="#internal-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Unexported functions for internal use</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="LinearOperator-class.html">LinearOperator-class</a></code> </p>
        </td>
        <td><p>Represent a sparse matrix-vector product operation</p></td>
      </tr><tr><td>
          <p><code><a href="linear_operator.html">linear_operator()</a></code> </p>
        </td>
        <td><p>Construct a LinearOperator object</p></td>
      </tr><tr><td>
          <p><code><a href="iterate_matrix.html">iterate_matrix()</a></code> </p>
        </td>
        <td><p>Get a wrapped pointer to the iterable matrix, in the form of an XPtrList (see fragments.R)</p></td>
      </tr><tr><td>
          <p><code><a href="matrix_type.html">matrix_type()</a></code> </p>
        </td>
        <td><p>Get the matrix data type (mat_uint32_t, mat_float, or mat_double for now)</p></td>
      </tr><tr><td>
          <p><code><a href="matrix_is_transform.html">matrix_is_transform()</a></code> </p>
        </td>
        <td><p>Return boolean whether the matrix is a transform</p></td>
      </tr><tr><td>
          <p><code><a href="merge_dimnames.html">merge_dimnames()</a></code> </p>
        </td>
        <td><p>Helper function for rbind/cbind merging dimnames</p></td>
      </tr><tr><td>
          <p><code><a href="concat_dimnames.html">concat_dimnames()</a></code> </p>
        </td>
        <td><p>Helper function for rbind/cbind concatenating dimnames</p></td>
      </tr><tr><td>
          <p><code><a href="normalized_dimnames.html">normalized_dimnames()</a></code> </p>
        </td>
        <td><p>Helper function to set dimnames to NULL instead of 0-length character vectors</p></td>
      </tr><tr><td>
          <p><code><a href="distanceToNearestDirectional.html">distanceToNearestDirectional()</a></code> </p>
        </td>
        <td><p>Find signed distance to nearest genomic ranges</p></td>
      </tr><tr><td>
          <p><code><a href="extendGenomicRanges.html">extendGenomicRanges()</a></code> </p>
        </td>
        <td><p>Extend GenomicRanges in a strand-aware fashion.</p></td>
      </tr><tr><td>
          <p><code><a href="fragments_identical.html">fragments_identical()</a></code> </p>
        </td>
        <td><p>Check if two fragments objects are identical</p></td>
      </tr><tr><td>
          <p><code><a href="XPtrList-class.html">XPtrList-class</a></code> </p>
        </td>
        <td><p>This class is used to hold a list of external pointers.</p></td>
      </tr><tr><td>
          <p><code><a href="wrapFragments.html">wrapFragments()</a></code> </p>
        </td>
        <td><p>Wrap an inner XPtrList with an outer externalptr object (Rcpp::XPtr)</p></td>
      </tr><tr><td>
          <p><code><a href="wrapMatrix.html">wrapMatrix()</a></code> </p>
        </td>
        <td><p>Construct an S4 matrix object wrapping another matrix object</p></td>
      </tr><tr><td>
          <p><code><a href="ensure_downloaded.html">ensure_downloaded()</a></code> </p>
        </td>
        <td><p>Download a file with a custom timeout</p></td>
      </tr><tr><td>
          <p><code><a href="normalize_ranges.html">normalize_ranges()</a></code> </p>
        </td>
        <td><p>Normalize an object representing genomic ranges</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Benjamin Parks.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

